docs: initial managed vault specs (#52)
noble-assets/dollar: spec/vaults-v2/00_overview.md

Reviewer: There is a huge discrepancy between when `USDN` is used, and when `$USDN` is used.

Author: made it all $USDN

> x`# Vaults V2 Overview

Reviewer: ```suggestion
# Managed Vault Overview
```

## Introduction

> The Noble Dollar Vault V2 features a single vault on the Noble chain that enables $USDN holders to participate in diversified yield opportunities by managing multiple remote positions across different protocols and chains through a professionally managed structure.

Reviewer: ```suggestion
This system features a single vault on the Noble chain that enables $USDN holders to participate in diversified yield opportunities by managing multiple remote positions across different protocols and chains through a professionally managed structure.
```

## Core Architecture

### Single Vault, Multiple Remote Positions

> The V2 system features a single vault on the Noble chain that can maintain multiple remote positions simultaneously across different chains. These remote positions are ERC-4626 compatible vaults. This architecture provides:

Reviewer: ```suggestion
Remote positions are [ERC-4626](https://eips.ethereum.org/EIPS/eip-4626) compatible vaults. This architecture provides:
```

- **Diversification**: Risk is spread across multiple yield sources managed by the single vault
- **Optimization**: The vault can dynamically allocate capital to the highest-performing strategies
- **Resilience**: Issues with one protocol don't compromise the entire vault
- **Flexibility**: New opportunities can be added without system migration

The Noble vault can configure:
- Maximum number of remote positions (e.g., 5-10)
- Allowed protocols (e.g., Hyperliquid, Base lending protocols, Noble AppLayer vaults)

> - Allowed chains (via Hyperlane Domain IDs: 998 for Hyperliquid, 8453 for Base, 1313817164 for Noble Mainnet)

Reviewer: ```suggestion
- Allowed chains (via Hyperlane Domain IDs: 998 for Hyperliquid, 8453 for Base, 1313817164 for Noble)
```

## Inflight Funds Management

### Overview

Inflight funds represent capital that is temporarily in transit between the Noble vault and its remote positions, or between positions during rebalancing. Each inflight transaction is tracked by its specific Hyperlane route identifier, allowing precise monitoring of capital flows across different bridge paths. This capital remains fully accounted for in the NAV to ensure accurate vault valuation at all times.

### Inflight Fund Types

1. **Deposit to Position**: USDN being deployed from the Noble vault to a remote ERC-4626 compatible vault
2. **Withdrawal from Position**: USDN returning from redeemed vault shares to the Noble vault
3. **Rebalance Between Positions**: USDN moving between remote vaults (via Noble after share redemption)
4. **Pending Deployment**: USDN from deposits awaiting allocation to remote vaults
5. **Pending Withdrawal Distribution**: USDN from redeemed shares awaiting distribution to withdrawal queue
6. **Yield Collection**: Automatic compounding within remote vaults

### Tracking Mechanism

Each inflight transaction maintains:

> - **Hyperlane Route ID**: Unique route identifier (e.g., 4000260998 for Noble→Hyperliquid)

Reviewer: Let's just remove the example here as Hyperlane routes are humanized as hex strings but underneath the hood are simple uint64's

```suggestion
- **Hyperlane Route ID**: Unique route identifier
```

- **Transaction ID**: Hyperlane message ID for the specific transfer

> - **Source/Destination Domains**: Hyperlane domain IDs for the route endpoints (1313817164 for Noble, 998 for Hyperliquid, 8453 for Base)

Reviewer: ```suggestion
- **Source/Destination Domains**: Hyperlane domain IDs for the route endpoints (998 for Hyperliquid, 8453 for Base, 1313817164 for Noble)
```

    ↓ Pass
Limit Checks → [Fail: Reject]
    ↓ Pass
Cooldown Check → [Fail: Reject]
    ↓ Pass
Calculate Shares (Current NAV)
    ↓
Mint Shares to User
    ↓
Update Velocity Metrics
    ↓
Mark Funds as Inflight
    ↓
Deploy Capital via Bridge
    ↓
Monitor Bridge Confirmation
    ↓
Deposit into Remote Vault
    ↓
Receive and Track Vault Shares
```

### Withdrawal Flow with Queue

```
Withdrawal Request
    ↓
Lock User Shares
    ↓
Enter Queue (FIFO)
    ↓
Record NAV Snapshot
    ↓
Wait for Liquidity
    ↓
Process Queue (Keeper/Auto)
    ↓
Mark as CLAIMABLE
    ↓
User Claims (After Delay)
    ↓
Burn Shares & Transfer $USDN
```

### Remote Position Capital Return Flow

```
Initiate Vault Share Redemption
    ↓
Redeem Shares for USDN
    ↓
Mark Expected USDN as Inflight
    ↓
NAV Includes Inflight USDN Value
    ↓
Hyperlane Confirmation Received
    ↓
Mark Transaction as Completed
    ↓
Mark as PENDING_WITHDRAWAL_DISTRIBUTION
    ↓
Update Vault Liquidity
    ↓
Process Withdrawal Queue
```

### Rebalancing Between Positions Flow

```
Initiate Rebalance Strategy
    ↓
Redeem Shares from Source Vault
    ↓
Mark as WITHDRAWAL_FROM_POSITION Inflight (Route A)
    ↓

> Track via Hyperlane Route ID (e.g., 998_4000260)

Reviewer: ```suggestion
Track via Hyperlane Route ID
```

    ↓
USDN Arrives at Noble
    ↓
Mark as PENDING_DEPLOYMENT
    ↓
Deploy to Target Vault
    ↓
Mark as DEPOSIT_TO_POSITION Inflight (Route B)
    ↓

> Track via Hyperlane Route ID (e.g., 4000260_8453)

Reviewer: ```suggestion
Track via Hyperlane Route ID
```

### Improved Security

- **Queue System**: Eliminates many attack vectors
- **Velocity Controls**: Prevents rapid manipulation
- **NAV Snapshots**: Fair value for all users
- **Oracle Redundancy**: Multiple verification sources

### Better User Experience

- **Single Token**: Users hold shares in the single Noble vault, complexity abstracted
- **Automatic Optimization**: No manual strategy switching
- **Transparent NAV**: Clear value per share
- **Predictable Withdrawals**: Queue provides certainty

## Emergency Procedures

### Pause Mechanisms

- **Deposit Pause**: Stop new deposits, allow withdrawals
- **Withdrawal Pause**: Emergency only, requires governance
- **Position Pause**: Halt specific remote positions
- **Global Pause**: Complete system halt (extreme cases)

### Recovery Procedures

- **Position Recovery**: Retrieve funds from failed positions
- **Inflight Recovery**: Manual intervention for stuck bridge transactions
- **Oracle Fallback**: Manual NAV updates if oracles fail
- **Transaction Override**: Force-complete stale inflight funds
- **Emergency Withdrawal**: Direct redemption at last known NAV
- **Governance Override**: Multi-sig can intervene if needed

> 
> ## Conclusion
> 
> The Noble Dollar Vault V2 system represents a sophisticated approach to decentralized yield optimization, balancing the need for high returns with robust security mechanisms. Through innovations like the withdrawal queue, the single vault's ability to manage multiple remote positions, and comprehensive anti-manipulation controls, the system provides users with professional-grade yield strategies while maintaining the security and fairness expected in DeFi.
> 
> The architecture's flexibility allows for future enhancements and new strategy additions without requiring user migration, ensuring the system can evolve with the rapidly changing DeFi landscape while protecting user value.

Reviewer: Personally I don't think a specification overview needs a conclusion 😅 

```suggestion
```

# Vaults V2 Overview

## Introduction

> The Noble Dollar Vault V2 system represents a significant evolution in decentralized yield optimization, introducing sophisticated mechanisms for managing multi-chain yield strategies while maintaining security and fair value distribution. This system features a single vault on the Noble chain that enables $USDN holders to participate in diversified yield opportunities by managing multiple remote positions across different protocols and chains through a professionally managed structure.

Author: Removing

## Core Architecture

### Single Vault, Multiple Remote Positions

The V2 system features a single vault on the Noble chain that can maintain multiple remote positions simultaneously across different chains. These remote positions are ERC-4626 compatible vaults. This architecture provides:

- **Diversification**: Risk is spread across multiple yield sources managed by the single vault
- **Optimization**: The vault can dynamically allocate capital to the highest-performing strategies
- **Resilience**: Issues with one protocol don't compromise the entire vault
- **Flexibility**: New opportunities can be added without system migration

The Noble vault can configure:
- Maximum number of remote positions (e.g., 5-10)

> - Allowed protocols (e.g., Hyperliquid, Base lending protocols, Noble App Layer vaults)

Author: Replaced

> - Allowed chains (via Hyperlane Domain IDs: 998 for Hyperliquid, 8453 for Base, 4000261 for Noble App Layer)

Reviewer: Also I'm not sure where this number comes from @zmanian, as we don't even have a Chain ID for the AppLayer yet.

Please review this and just list Hyperliquid and Base!

   - Maturing positions
   - Incoming deposits
4. **Fulfillment**: When liquidity available, requests are marked CLAIMABLE
5. **Claim Phase**: Users claim their $USDN after withdrawal delay period

#### Fair Value Protection

- Each request uses the NAV at time of request, not fulfillment
- Prevents front-running of NAV updates
- Eliminates sandwich attacks around large deposits/withdrawals
- Ensures all users receive fair value regardless of queue position

## Security Mechanisms

### Anti-Manipulation Framework

The system implements multiple layers of defense against malicious behavior:

#### 1. Deposit Velocity Controls

```
Velocity Score = (Recent Volume / Time Window) × (Deposit Count / Expected Count)
```

- Tracks deposit patterns over rolling time windows
- Flags suspicious rapid deposits that could indicate attacks
- Enforces cooldown periods for high-velocity users

#### 2. Deposit Limits

Multiple limit types work in concert:

- **Per-User Limits**: Maximum total deposit per address
- **Per-Block Limits**: Maximum deposits in single block
- **Per-Transaction Limits**: Maximum single deposit amount
- **Total Vault Limits**: Maximum total deposits across all users

> - **Minimum Amounts**: Prevents dust attacks and griefing

Author: My mental model is that withdrawals are very light weight because they only trigger NAV accounting when they are serviced by the operator so that the risk of spam is low and min_size would make it unlikely that anyone would be able to grief the chain.

#### 3. Share Price Manipulation Defense

- **Entry/Exit Delays**: Withdrawal queue prevents immediate round-trips
- **NAV Snapshots**: Lock in values at request time
- **Slippage Protection**: Maximum acceptable NAV deviation checks
- **Fee Structure**: Management and performance fees discourage short-term speculation

#### 4. Oracle Security

Remote position values are protected through:

- **Staleness Checks**: Reject updates older than configured threshold

> - **Deviation Limits**: Flag suspicious large value changes

Author: There are two options here.

1. An extremely large change in NAV from the oracle could automatically trigger emergency mode. Basically stop deposits and withdrawals. The operator could intervene if this is normal. 

2. Just log the large change in NAV and have the operator step in to trigger emergency mode.

Remote positions leverage Hyperlane for secure cross-chain operations:

- **Deployment**: USDN bridged via specific Hyperlane routes to deposit into ERC-4626 compatible vaults
- **Share Management**: Remote shares received and tracked for each remote position
- **Inflight Tracking**: USDN marked as inflight per Hyperlane route ID during bridge transit
- **Route Management**: Each Hyperlane route (e.g., Noble→Hyperliquid, Base→Noble) tracked separately
- **Push-Based Price Updates**: Remote chains proactively push share prices to Noble via Hyperlane using fixed-length byte encoding for efficiency
- **Automatic Value Updates**: Noble continuously receives and applies byte-encoded price data directly from the Hyperlane Mailbox
- **Redemptions**: Remote vault shares redeemed for USDN and bridged back to Noble via specific return routes
- **Completion Tracking**: Monitoring of bridge transaction completions per route
- **Emergency Recovery**: Fallback mechanisms for bridge failures on specific routes

### Risk Management

Each remote position is subject to:

- **Concentration Limits**: No single position > X% of total vault
- **Chain Limits**: Maximum exposure per blockchain

> - **Approved Vaults Only**: Only deploy to pre-approved vault addresses

Author: Does the Noble team have a preference for who gets to approve the deployment addresses?

In most systems we've built we have had a dual control mechanism where the customer or another party is involved in decisions to add or remove targe positions.

I think it's a good pattern but NASD or  Noble authority could work here.

### Inflight Fund Types

1. **Deposit to Position**: USDN being deployed from the Noble vault to a remote ERC-4626 compatible vault
2. **Withdrawal from Position**: USDN returning from redeemed remote vault shares to the Noble vault
3. **Rebalance Between Positions**: USDN moving between remote vaults (via Noble after share redemption)
4. **Pending Deployment**: USDN from deposits awaiting allocation to remote vaults
5. **Pending Withdrawal Distribution**: USDN from redeemed shares awaiting distribution to withdrawal queue
6. **Yield Collection**: Automatic compounding within remote vaults

### Tracking Mechanism

Each inflight transaction maintains:
- **Hyperlane Route ID**: Unique route identifier (e.g., 4000260998 for Noble→Hyperliquid)
- **Transaction ID**: Hyperlane message ID for the specific transfer

> - **Source/Destination Domains**: Hyperlane domain IDs for the route endpoints (4000260 for Noble, 998 for Hyperliquid, 8453 for Base, 4000261 for Noble App Layer)

Reviewer: https://github.com/noble-assets/dollar/pull/52#discussion_r2297435410

- **Expected Value**: USDN amount sent including estimated bridge fees
- **Current Value**: Last known USDN value for NAV calculation
- **Time Bounds**: Expected arrival time and maximum duration per route
- **Status Updates**: PENDING → CONFIRMED → COMPLETED lifecycle per route
- **Route-Specific Limits**: Maximum exposure allowed per Hyperlane route

### NAV Impact

Inflight funds are included in NAV calculations to prevent artificial value fluctuations:

```
During Transit:
- USDN leaves source → Marked as inflight
- NAV unchanged (USDN still counted)
- Hyperlane confirms → Status: CONFIRMED
- USDN arrives → Status: COMPLETED

Special States:
- New deposits → PENDING_DEPLOYMENT until allocated
- Returned funds → PENDING_WITHDRAWAL_DISTRIBUTION until claimed
- Rebalancing → WITHDRAWAL then PENDING_DEPLOYMENT then DEPOSIT
```

### Transaction Completion

1. **Status Tracking**: Bridge transactions monitored for completion
2. **Timeout Management**: Stale transactions flagged for investigation

> 3. **Manual Intervention**: Failed transactions require governance action

Author: Hyperlane transactions don't transport layer timeouts. Failures basically an upprogresses state machine over a generous amount of time.

### Risk Mitigation

- **Maximum Duration Limits**: Funds cannot remain inflight indefinitely on any route
- **Per-Route Value Caps**: Limits on inflight exposure for each Hyperlane route
- **Total Value Caps**: Aggregate limits across all routes for the vault
- **Route Diversification**: Use multiple Hyperlane routes to reduce concentration risk

> - **Insurance Reserve**: Coverage for potential failures on specific routes

Author: Just gonna take this out.

    ↓
User Claims (After Delay)
    ↓
Burn Shares & Transfer $USDN
```

### Remote Position Capital Return Flow

```
Initiate Vault Share Redemption
    ↓
Redeem Shares for USDN
    ↓
Mark Expected USDN as Inflight
    ↓
NAV Includes Inflight USDN Value
    ↓
Hyperlane Confirmation Received
    ↓
Mark Transaction as Completed
    ↓
Mark as PENDING_WITHDRAWAL_DISTRIBUTION
    ↓
Update Vault Liquidity
    ↓
Process Withdrawal Queue
```

### Rebalancing Between Positions Flow

```
Initiate Rebalance Strategy
    ↓
Redeem Shares from Source Vault
    ↓
Mark as WITHDRAWAL_FROM_POSITION Inflight (Route A)
    ↓
Track via Hyperlane Route ID (e.g., 998_4000260)
    ↓
USDN Arrives at Noble
    ↓
Mark as PENDING_DEPLOYMENT
    ↓
Deploy to Target Vault
    ↓
Mark as DEPOSIT_TO_POSITION Inflight (Route B)
    ↓
Track via Hyperlane Route ID (e.g., 4000260_8453)
    ↓
Deposit into Target Vault & Receive Remote Position Shares
    ↓
Update Share Balances & Clear Route Tracking
```

## Economic Model

### Fee Structure

- **Management Fee**: Annual percentage on total AUM
- **Performance Fee**: Percentage of profits above hurdle rate
- **No Deposit/Withdrawal Fees**: Encourages participation

> - **Fee Distribution**: To protocol treasury and/or stakers

Author: This could be to noble token holders. I think the " where to do fees" go question in still be resolved.
